hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java

```diff
@@ -3215,7 +3215,13 @@ else if ( discriminatorValue == NOT_NULL_DISCRIMINATOR ) {
 				return junction;
 			}
 
-			junction.add( new NullnessPredicate( sqlExpression ) );
+			if ( hasNonNull ) {
+				junction.add( new NegatedPredicate( new NullnessPredicate( sqlExpression ) ) );			
```

Comment:
Should return `junction` to eliminate unnecessary `InListPredicate` since `DTYPE in (...)` is subset of `DTYPE is not null`.